src/org/rascalmpl/util/maven/Artifact.java

```diff
@@ -181,109 +181,108 @@ public boolean equals(Object obj) {
+                    // - if a provided dependency is present in the maven repository it's considered "provided"
+                    var pomDep = artifact.ourResolver.calculatePomPath(d.getCoordinate());
+                    if (!Files.notExists(pomDep)) {
+                        // ok, doesn't exist yet. so don't download it to calculate dependencies
+                        // and don't add it to the list since somewhere else somebody might depend on it
+                        continue;
+                    }
                 }
-            }
-
-            String resolvedVersion = resolvedVersions.get(versionLess);
-            if (resolvedVersion != null) {
-                alreadyResolved = true;
-                coordinate = new ArtifactCoordinate(coordinate.getGroupId(), coordinate.getArtifactId(), resolvedVersion, coordinate.getClassifier());
-            } else {
-                alreadyResolved = false;
-                resolvedVersions.put(versionLess, version);
-            }
-
-            var key = new ResolveKey(coordinate, excludes);
-            if (alreadySolved.contains(key) || !d.shouldInclude(forScope)) {
-                continue;
-            }
-            alreadySolved.add(key);
-
-            if (d.getScope() == Scope.PROVIDED) {
-                // current maven behavior seems to be:
-                // - do not download provided dependencies
-                // - if a provided dependency is present in the maven repository it's considered "provided"
-                var pomDep = ourResolver.calculatePomPath(d.getCoordinate());
-                if (!Files.notExists(pomDep)) {
-                    // ok, doesn't exist yet. so don't download it to calculate dependencies
-                    // and don't add it to the list since somewhere else somebody might depend on it
+                if (d.getScope() == Scope.SYSTEM) {
+                    if (!resolved) {
+                        result.add(createSystemArtifact(d));
+                    }
                     continue;
                 }
-            }
-            if (d.getScope() == Scope.SYSTEM) {
-                if (!alreadyResolved) {
-                    result.add(createSystemArtifact(d));
-                }
-                continue;
-            }
-
-            Set<WithoutVersion> newExcludes = excludes;
-            if (!d.getExclusions().isEmpty()) {
-                newExcludes = new HashSet<>(excludes);
-                newExcludes.addAll(d.getExclusions());
-            }
-            newExcludes = Collections.unmodifiableSet(newExcludes);
 
-            var art = parser.parseArtifact(coordinate, newExcludes, ourResolver);
-            if (art != null) {
-                if (!alreadyResolved) {
-                    result.add(art);
+                Set<WithoutVersion> newExclusions = state.exclusions;
+                if (!d.getExclusions().isEmpty()) {
+                    newExclusions = new HashSet<>(state.exclusions);
```

Comment:
- nit: if `state.exclusions` is empty, we can instead say that `newExclusions = d.getExclusions()` and skip this copying mess.
- nit2: instead of repeating `state.exclusions` we should be reusing `newExclusions` here, since that is the one we're growing.